kobo-uncaged/kunc/kunc.go

```diff
@@ -0,0 +1,296 @@
+	"github.com/shermp/Kobo-UNCaGED/kobo-uncaged/util"
+	"github.com/shermp/UNCaGED/uc"
+)
+
+type koboUncaged struct {
+	k *device.Kobo
+}
+
+// New initialises the koboUncaged object that will be passed to UNCaGED
+func New(kobo *device.Kobo) *koboUncaged {
+	ku := &koboUncaged{}
+	ku.k = kobo
+	return ku
+}
+
+// GetClientOptions returns all the client specific options required for UNCaGED
+func (ku *koboUncaged) GetClientOptions() uc.ClientOptions {
+	opts := uc.ClientOptions{}
+	opts.ClientName = "Kobo UNCaGED " // + kuVersion
+	var ext []string
+	var thumbSz image.Point
+	ext, opts.DeviceModel, thumbSz = ku.k.GetDeviceOptions()
+	opts.SupportedExt = append(opts.SupportedExt, ext...)
+	opts.DeviceName = "Kobo"
+	opts.CoverDims.Width, opts.CoverDims.Height = thumbSz.X, thumbSz.Y
+	return opts
+}
+
+// GetDeviceBookList returns a slice of all the books currently on the device
+// A nil slice is interpreted has having no books on the device
+func (ku *koboUncaged) GetDeviceBookList() []uc.BookCountDetails {
+	bc := []uc.BookCountDetails{}
+	for _, md := range ku.k.MetadataMap {
+		lastMod := time.Now()
+		if md.LastModified != nil {
+			lastMod, _ = time.Parse(time.RFC3339, *md.LastModified)
+		}
+		bcd := uc.BookCountDetails{
+			UUID:         md.UUID,
+			Lpath:        md.Lpath,
+			LastModified: lastMod,
+		}
+		bcd.Extension = filepath.Ext(md.Lpath)
+		bc = append(bc, bcd)
+	}
+	//spew.Dump(bc)
+	return bc
+}
+
+// GetMetadataList sends complete metadata for the books listed in lpaths, or for
+// all books on device if lpaths is empty
+func (ku *koboUncaged) GetMetadataList(books []uc.BookID) []map[string]interface{} {
+	//spew.Dump(ku.k.MetadataMap)
+	//spew.Dump(books)
+	mdList := []map[string]interface{}{}
+	if len(books) > 0 {
+		for _, bk := range books {
+			cid := util.LpathToContentID(bk.Lpath, string(ku.k.ContentIDprefix))
+			fmt.Println(cid)
+			md := map[string]interface{}{}
+			//spew.Dump(ku.k.MetadataMap[cid])
+			mapstructure.Decode(ku.k.MetadataMap[cid], &md)
+			mdList = append(mdList, md)
+		}
+	} else {
+		for _, kmd := range ku.k.MetadataMap {
+			md := map[string]interface{}{}
+			//spew.Dump(kmd)
+			mapstructure.Decode(kmd, &md)
+			mdList = append(mdList, md)
+		}
+	}
+	return mdList
+}
+
+// GetDeviceInfo asks the client for information about the drive info to use
+func (ku *koboUncaged) GetDeviceInfo() uc.DeviceInfo {
+	return ku.k.DriveInfo
+}
+
+// SetDeviceInfo sets the new device info, as comes from calibre. Only the nested
+// struct DevInfo is modified.
+func (ku *koboUncaged) SetDeviceInfo(devInfo uc.DeviceInfo) {
+	ku.k.DriveInfo = devInfo
+	ku.k.SaveDeviceInfo()
+}
+
+// UpdateMetadata instructs the client to update their metadata according to the
+// new slice of metadata maps
+func (ku *koboUncaged) UpdateMetadata(mdList []map[string]interface{}) {
+	for _, md := range mdList {
+		koboMD := device.CreateKoboMetadata()
+		mapstructure.Decode(md, &koboMD)
```

Comment:
```suggestion
		if err := mapstructure.Decode(md, &koboMD); err != nil {
			panic(err)
		}
```

The errors returned by decode all seem to be programming errors, and should probably not be silently ignored.

Reply:
The metadata map list ultimately comes directly from calibre. Not sure if I want to be panicing here.

Reply:
I'd rather not fail silently though.

Reply:
It ultimately comes down to an interface issue again. I either log it, or do nothing at the moment.